storage/src/lib.rs

```diff
@@ -410,22 +410,70 @@ impl Storage {
     /// - Ok of Some entry - if found in loose index
     /// - Ok of None - if requested height is low enough to not be in the index already
     /// - Err of `BlockHeightNotFound` - if requested height is too large and out of index bounds
+    /// NOTE: this function does NOT return EBB, even if it's a part of the loose index
     pub fn get_from_loose_index(
         &self,
         height: ChainDifficulty,
     ) -> Result<Option<&LooseChainHeightEntry>> {
-        if let Some(ref idx_tip) = self.chain_height_idx.loose_idx.first() {
-            let find_diff = u64::from(height);
-            let tip_diff = u64::from(idx_tip.difficulty);
-            if find_diff > tip_diff {
-                return Err(Error::BlockHeightNotFound(find_diff));
+        match self.chain_height_idx.loose_idx.first() {
+            None => {
+                // No loose index is available at the moment at all
+                return Ok(None);
             }
-            let idx = (tip_diff - find_diff) as usize;
-            if self.chain_height_idx.loose_idx.len() > idx {
-                return Ok(Some(&(self.chain_height_idx.loose_idx[idx])));
+            Some(ref idx_tip) => {
+                // idx_tip is the NEWEST loose block
+                let find_diff = u64::from(height);
+                let tip_diff = u64::from(idx_tip.difficulty);
+                if find_diff > tip_diff {
+                    return Err(Error::BlockHeightNotFound(find_diff));
+                }
+                let idx = (tip_diff - find_diff) as usize;
+                let loose_index_len = self.chain_height_idx.loose_idx.len();
+                if loose_index_len <= idx {
+                    // We are searching deeper in history than the loose index
+                    return Ok(None);
+                }
+                let entry = &self.chain_height_idx.loose_idx[idx];
+                if !entry.date.is_boundary() {
+                    return Ok(Some(entry));
+                }
+                debug!("Search by height in loose index found EBB. Proceeding with caution.");
+                if (loose_index_len - idx) <= 1 {
+                    debug!("EBB is the oldest available loose entry. Returning nothing, to continue search in historical data.");
+                    return Ok(None);
+                }
+                // There are loose blocks older than found EBB, trying next one
+                let idx2 = idx + 1;
+                let entry2 = &self.chain_height_idx.loose_idx[idx2];
+                // Validating next found entry
+                let error_msg: Option<&str> =
+                    if u64::from(entry2.difficulty) != find_diff {
+                        Some("the height does not match")
+                    } else if entry2.date.is_boundary() {
+                        Some("it's also an EBB")
+                    } else {
+                        None
+                    };
+                if let Some(error_msg) = error_msg {
```

Comment:
Is this the place we want to do this validation? Should we not do some kind of structure validation upon building/modification of the loose index so we can notice it earlier?

Reply:
Yeah, I mostly added it as a result of paranoia attack 😄Kinda like if anything goes wrong then at least we get the error logs. We actually have the height continuity validation when new entries are added (https://github.com/Emurgo/rust-cardano/blob/8065f310dad6823b7c9319008260bb2702c7bbbd/storage/src/lib.rs#L351-L367) so maybe I can just add the sanity check for two consecutive EBBs there too.

But when I think about it, I think that this check also makes sense in the getter function too, because the main requirement for this function is that if it returns some block, then it should be NOT an EBB and `block.difficutly` MUST match the function argument (height). So this par is just kinda makes it extra-sure that this function will not break the contract, and in the worst case will at least log the constant 🤔

Reply:
I guess it doesn't hurt to have it in here if we already have the checks sans the consecutive EBB in the entry adding. It only runs when we are already in a weird case anyway, so the overhead is basically non-existent, and would also serve as extra protection in case the entry adding code gets changed or added to and somehow something slips through.